// @rn-primitives/slot/dist/index.mjs
'use client'

// src/slot.tsx
import * as React from "react";
function Slot(props) {
  const { children, ref: forwardedRef, ...restOfProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot - Invalid asChild element", children);
    return null;
  }
  if (isTextChildren(children)) {
    console.log("Slot - Text children are not supported", children);
    return null;
  }
  const childrenProps = children.props ?? {};
  if (children.type === React.Fragment) {
    return <>
        {React.Children.toArray(childrenProps.children).map(
      (child) => React.isValidElement(child) ? Slot({ ...restOfProps, ref: forwardedRef, children: child }) : child
    )}
      </>;
  }
  const { ref: childRef, ...childProps } = childrenProps;
  return React.cloneElement(children, {
    ...mergeProps(restOfProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
}
Slot.displayName = "Slot";
var Pressable = ({
  ref: forwardedRef,
  ...props
}) => {
  const { children, ...pressableSlotProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot.Pressable - Invalid asChild element", children);
    return null;
  }
  const childProps = children.props ?? {};
  const childRef = childProps.ref;
  return React.cloneElement(isTextChildren(children) ? <></> : children, {
    ...mergeProps(pressableSlotProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
};
Pressable.displayName = "SlotPressable";
var View = ({
  ref: forwardedRef,
  ...props
}) => {
  const { children, ...viewSlotProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot.View - Invalid asChild element", children);
    return null;
  }
  const childProps = children.props ?? {};
  const childRef = childProps.ref;
  return React.cloneElement(isTextChildren(children) ? <></> : children, {
    ...mergeProps(viewSlotProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
};
View.displayName = "SlotView";
var Text = ({
  ref: forwardedRef,
  ...props
}) => {
  const { children, ...textSlotProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot.Text - Invalid asChild element", children);
    return null;
  }
  const childProps = children.props ?? {};
  const childRef = childProps.ref;
  return React.cloneElement(isTextChildren(children) ? <></> : children, {
    ...mergeProps(textSlotProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
};
Text.displayName = "SlotText";
var Image = ({
  ref: forwardedRef,
  ...props
}) => {
  const { children, ...imageSlotProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot.Image - Invalid asChild element", children);
    return null;
  }
  const childProps = children.props ?? {};
  const childRef = childProps.ref;
  return React.cloneElement(isTextChildren(children) ? <></> : children, {
    ...mergeProps(imageSlotProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
};
Image.displayName = "SlotImage";
function setRef(ref, value) {
  if (typeof ref === "function") {
    const cleanup = ref(value);
    if (typeof cleanup === "function") {
      return cleanup;
    }
    return;
  }
  if (ref != null) {
    ref.current = value;
    return () => {
      ref.current = null;
    };
  }
}
function composeRefs(...refs) {
  let cleanups = [];
  return (node) => {
    cleanups.forEach((cleanup) => cleanup());
    cleanups = [];
    if (node == null) {
      refs.forEach((ref) => {
        if (typeof ref === "function") {
          ref(null);
        } else if (ref != null) {
          ref.current = null;
        }
      });
      return;
    }
    cleanups = refs.map((ref) => setRef(ref, node)).filter((cleanup) => cleanup != null);
  };
}
function mergeProps(slotProps, childProps) {
  const overrideProps = { ...childProps };
  for (const propName in childProps) {
    const slotPropValue = slotProps[propName];
    const childPropValue = childProps[propName];
    const isHandler = /^on[A-Z]/.test(propName);
    if (isHandler) {
      if (slotPropValue && childPropValue) {
        overrideProps[propName] = (...args) => {
          childPropValue(...args);
          slotPropValue(...args);
        };
      } else if (slotPropValue) {
        overrideProps[propName] = slotPropValue;
      }
    } else if (propName === "style") {
      overrideProps[propName] = combineStyles(slotPropValue, childPropValue);
    } else if (propName === "className") {
      overrideProps[propName] = [slotPropValue, childPropValue].filter(Boolean).join(" ");
    }
  }
  return { ...slotProps, ...overrideProps };
}
function combineStyles(slotStyle, childValue) {
  if (typeof slotStyle === "function" && typeof childValue === "function") {
    return (state) => {
      return [slotStyle(state), childValue(state)];
    };
  }
  if (typeof slotStyle === "function") {
    return (state) => {
      return childValue ? [slotStyle(state), childValue] : slotStyle(state);
    };
  }
  if (typeof childValue === "function") {
    return (state) => {
      return slotStyle ? [slotStyle, childValue(state)] : childValue(state);
    };
  }
  if (slotStyle && childValue) {
    return [slotStyle, childValue];
  }
  return slotStyle ?? childValue;
}
function isTextChildren(children) {
  return Array.isArray(children) ? children.every((child) => typeof child === "string") : typeof children === "string";
}
export {
  Image,
  Pressable,
  Slot,
  Text,
  View,
  isTextChildren
};
