// @rn-primitives/slot/dist/index.js
'use client'
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Image: () => Image,
  Pressable: () => Pressable,
  Slot: () => Slot,
  Text: () => Text,
  View: () => View,
  isTextChildren: () => isTextChildren
});
module.exports = __toCommonJS(src_exports);

// src/slot.tsx
var React = __toESM(require("react"));
function Slot(props) {
  const { children, ref: forwardedRef, ...restOfProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot - Invalid asChild element", children);
    return null;
  }
  if (isTextChildren(children)) {
    console.log("Slot - Text children are not supported", children);
    return null;
  }
  const childrenProps = children.props ?? {};
  if (children.type === React.Fragment) {
    return <>
        {React.Children.toArray(childrenProps.children).map(
      (child) => React.isValidElement(child) ? Slot({ ...restOfProps, ref: forwardedRef, children: child }) : child
    )}
      </>;
  }
  const { ref: childRef, ...childProps } = childrenProps;
  return React.cloneElement(children, {
    ...mergeProps(restOfProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
}
Slot.displayName = "Slot";
var Pressable = ({
  ref: forwardedRef,
  ...props
}) => {
  const { children, ...pressableSlotProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot.Pressable - Invalid asChild element", children);
    return null;
  }
  const childProps = children.props ?? {};
  const childRef = childProps.ref;
  return React.cloneElement(isTextChildren(children) ? <></> : children, {
    ...mergeProps(pressableSlotProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
};
Pressable.displayName = "SlotPressable";
var View = ({
  ref: forwardedRef,
  ...props
}) => {
  const { children, ...viewSlotProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot.View - Invalid asChild element", children);
    return null;
  }
  const childProps = children.props ?? {};
  const childRef = childProps.ref;
  return React.cloneElement(isTextChildren(children) ? <></> : children, {
    ...mergeProps(viewSlotProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
};
View.displayName = "SlotView";
var Text = ({
  ref: forwardedRef,
  ...props
}) => {
  const { children, ...textSlotProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot.Text - Invalid asChild element", children);
    return null;
  }
  const childProps = children.props ?? {};
  const childRef = childProps.ref;
  return React.cloneElement(isTextChildren(children) ? <></> : children, {
    ...mergeProps(textSlotProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
};
Text.displayName = "SlotText";
var Image = ({
  ref: forwardedRef,
  ...props
}) => {
  const { children, ...imageSlotProps } = props;
  if (!React.isValidElement(children)) {
    console.log("Slot.Image - Invalid asChild element", children);
    return null;
  }
  const childProps = children.props ?? {};
  const childRef = childProps.ref;
  return React.cloneElement(isTextChildren(children) ? <></> : children, {
    ...mergeProps(imageSlotProps, childProps),
    ref: forwardedRef ? composeRefs(forwardedRef, childRef) : childRef
  });
};
Image.displayName = "SlotImage";
function setRef(ref, value) {
  if (typeof ref === "function") {
    const cleanup = ref(value);
    if (typeof cleanup === "function") {
      return cleanup;
    }
    return;
  }
  if (ref != null) {
    ref.current = value;
    return () => {
      ref.current = null;
    };
  }
}
function composeRefs(...refs) {
  let cleanups = [];
  return (node) => {
    cleanups.forEach((cleanup) => cleanup());
    cleanups = [];
    if (node == null) {
      refs.forEach((ref) => {
        if (typeof ref === "function") {
          ref(null);
        } else if (ref != null) {
          ref.current = null;
        }
      });
      return;
    }
    cleanups = refs.map((ref) => setRef(ref, node)).filter((cleanup) => cleanup != null);
  };
}
function mergeProps(slotProps, childProps) {
  const overrideProps = { ...childProps };
  for (const propName in childProps) {
    const slotPropValue = slotProps[propName];
    const childPropValue = childProps[propName];
    const isHandler = /^on[A-Z]/.test(propName);
    if (isHandler) {
      if (slotPropValue && childPropValue) {
        overrideProps[propName] = (...args) => {
          childPropValue(...args);
          slotPropValue(...args);
        };
      } else if (slotPropValue) {
        overrideProps[propName] = slotPropValue;
      }
    } else if (propName === "style") {
      overrideProps[propName] = combineStyles(slotPropValue, childPropValue);
    } else if (propName === "className") {
      overrideProps[propName] = [slotPropValue, childPropValue].filter(Boolean).join(" ");
    }
  }
  return { ...slotProps, ...overrideProps };
}
function combineStyles(slotStyle, childValue) {
  if (typeof slotStyle === "function" && typeof childValue === "function") {
    return (state) => {
      return [slotStyle(state), childValue(state)];
    };
  }
  if (typeof slotStyle === "function") {
    return (state) => {
      return childValue ? [slotStyle(state), childValue] : slotStyle(state);
    };
  }
  if (typeof childValue === "function") {
    return (state) => {
      return slotStyle ? [slotStyle, childValue(state)] : childValue(state);
    };
  }
  if (slotStyle && childValue) {
    return [slotStyle, childValue];
  }
  return slotStyle ?? childValue;
}
function isTextChildren(children) {
  return Array.isArray(children) ? children.every((child) => typeof child === "string") : typeof children === "string";
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Image,
  Pressable,
  Slot,
  Text,
  View,
  isTextChildren
});
