// @rn-primitives/progress/dist/progress.js
'use client'
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/progress.tsx
var progress_exports = {};
__export(progress_exports, {
  Indicator: () => Indicator,
  Root: () => Root
});
module.exports = __toCommonJS(progress_exports);
var import_slot = require("@rn-primitives/slot");
var import_react_native = require("react-native");
var DEFAULT_MAX = 100;
var Root = ({
  asChild,
  value: valueProp,
  max: maxProp,
  getValueLabel = defaultGetValueLabel,
  ref,
  ...props
}) => {
  const max = maxProp ?? DEFAULT_MAX;
  const value = isValidValueNumber(valueProp, max) ? valueProp : 0;
  const Component = asChild ? import_slot.Slot : import_react_native.View;
  return <Component
    role='progressbar'
    ref={ref}
    aria-valuemax={max}
    aria-valuemin={0}
    aria-valuenow={value}
    aria-valuetext={getValueLabel(value, max)}
    accessibilityValue={{
      min: 0,
      max,
      now: value,
      text: getValueLabel(value, max)
    }}
    {...props}
  />;
};
Root.displayName = "RootProgress";
var Indicator = ({ asChild, ref, ...props }) => {
  const Component = asChild ? import_slot.Slot : import_react_native.View;
  return <Component ref={ref} role='presentation' {...props} />;
};
Indicator.displayName = "IndicatorProgress";
function defaultGetValueLabel(value, max) {
  return `${Math.round(value / max * 100)}%`;
}
function isValidValueNumber(value, max) {
  return typeof value === "number" && !isNaN(value) && value <= max && value >= 0;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Indicator,
  Root
});
