// @rn-primitives/progress/dist/progress.mjs
'use client'

// src/progress.tsx
import { Slot } from "@rn-primitives/slot";
import { View } from "react-native";
var DEFAULT_MAX = 100;
var Root = ({
  asChild,
  value: valueProp,
  max: maxProp,
  getValueLabel = defaultGetValueLabel,
  ref,
  ...props
}) => {
  const max = maxProp ?? DEFAULT_MAX;
  const value = isValidValueNumber(valueProp, max) ? valueProp : 0;
  const Component = asChild ? Slot : View;
  return <Component
    role='progressbar'
    ref={ref}
    aria-valuemax={max}
    aria-valuemin={0}
    aria-valuenow={value}
    aria-valuetext={getValueLabel(value, max)}
    accessibilityValue={{
      min: 0,
      max,
      now: value,
      text: getValueLabel(value, max)
    }}
    {...props}
  />;
};
Root.displayName = "RootProgress";
var Indicator = ({ asChild, ref, ...props }) => {
  const Component = asChild ? Slot : View;
  return <Component ref={ref} role='presentation' {...props} />;
};
Indicator.displayName = "IndicatorProgress";
function defaultGetValueLabel(value, max) {
  return `${Math.round(value / max * 100)}%`;
}
function isValidValueNumber(value, max) {
  return typeof value === "number" && !isNaN(value) && value <= max && value >= 0;
}
export {
  Indicator,
  Root
};
